modules/websocket_bridge/README.md

```diff
@@ -0,0 +1,147 @@
+# Module - WebSocket Bridge
+
+## Overview
+
+This module offers a websocket interface to hephaestus, enabling foxglove studio or a similar websocket-capable client to connect to hephaestus communication and:
+  - Publish and subscribe to hephaestus topics
+  - Trigger hephaestus service calls
+  - Monitor the hephaestus node graph (Foxglove has a built in node graph visualizer)
+
+## Build
+
+To build the WebSocket bridge, execute the following command:
+
+```bash
+bazel build //modules/websocket_bridge:app
+```
+
+## Running the WebSocket Bridge
+
+To run the WebSocket bridge, execute the following command:
+
+```bash
+bazel run //modules/websocket_bridge::app -- -c <path to config file>
+```
+
+## Bridge Structure
+
+```mermaid
+classDiagram
+class WsBridge {
+    - WsBridgeConfig
+    - WsBridgeState
+    - IpcGraph
+    - IpcInterface
+
+    - Foxglove WebSocket Server
+    
+    + start()
+    + stop()
+
+    + callback__Ws__*()
+    + callback_IpcGraph_*()
+    + callback_Ipc_*()
+}
+
+class WsBridgeConfig {
+    - WS Server Config
+    - IPC Config
+    - Topic/Service Whitelists/Blacklists
+}
+
+class WsBridgeState {
+   - Map: IPC Topics <-> WS Channels
+   - Map: WS Channels <-> WS Clients
+   - Map: IPC Services <-> WS Services
+   - Map: WS Services <-> WS Clients
+
+   + add/remove/has...()
+}
+
+class IpcGraph {
+    - Topic/Type Database
+    - Discovery
+
+    + start()
+    + stop()
+
+    + callback_EndPointInfoUpdate()
+
+    + add/removePublisherEndpoint()
+    + add/removeSubscriberEndpoint()
+    + topicHasAnyEndpoint()
+    + add/remove/hasTopic()
+
+    + add/remove/hasServiceServerEndpoint()
+    + add/remove/hasServiceClientEndpoint()
+    + add/remove/hasService()
+}
+
+class IpcInterface {
+    - Subscribers
+    - Publishers
+    - ServiceClients
+
+    + start()
+    + stop()
+
+    + add/remove/hasSubscriber()
+    + add/remove/hasPublisher()
+
+    + callService()
+    + callServiceAsync()
+
+    + callback__ServiceResponse()
+}
+
+class Foxglove WebSocket Server {
+    - State/Clients
+    - Config
+
+    + start()
+    + stop()
+
+    + add/removeChannels()    
+    + sendMessage()
+
+    + add/removeServices()    
+    + sendServiceResponse()
+    + sendServiceError()
+
+    + updateConnectionGraph()    
+}
+
+WsBridge --> WsBridgeConfig
+WsBridge --> WsBridgeState
+WsBridge --> IpcGraph
+WsBridge --> IpcInterface
+WsBridge --> Foxglove WebSocket Server
+```
+
+## Test Clients
+
+### Test Client - Services - C++
```

Comment:
Do I understand correctly that this client is used to test the bridge instead of running foxglove studio? In that case this would be a server, not a client, no? Or am I missing something?

Reply:
I (and foxglove) considered the bridge the ws server. So these apps implement alternate ws clients purely in C++ to test the bridge without foxglove studio.

I had to implement them because foxglove studio does not currently support calling service calls via the foxglove studio GUI if the encoding is protobuf. Same for client-side topics.

On the bright side: These serve as half-decent c++ examples for writing additional clients that make use of the bridge, like maybe some development or production tools that you want talking to the robot from the outside without hooking them into the zenoh network.

Does this help?